libnetwork/drivers/bridge/bridge_linux.go

```diff
@@ -1078,27 +1078,29 @@ func (d *driver) CreateEndpoint(nid, eid string, ifInfo driverapi.InterfaceInfo,
 		return fmt.Errorf("could not set link up for host interface %s: %v", hostIfName, err)
 	}
 
+	// Generate a MAC-address-based IPv6 address, which may be used by the default
+	// bridge network instead of an IPAM allocated address.
 	if endpoint.addrv6 == nil && config.EnableIPv6 {
 		var ip6 net.IP
 		network := n.bridge.bridgeIPv6
 		if config.AddressIPv6 != nil {
 			network = config.AddressIPv6
 		}
+		// If there are fewer than 48 host-bits in the network, it's not possible to
+		// generate an IPv6 address - but the network size has already been checked if a
+		// MAC-based address is required.
```

Comment:
That wouldn't work, because the driver isn't told to allocate an IPv6 address - it just does it if there isn't already an address and the network is IPv6-enabled. It's more legacy-networking / default-bridge fun ...

In `Network.createEndpoint()`, if the daemon has asked for `NetworkOptionDeferIPv6Alloc(true)`, `n.postIPv6` is set. Then, its first call to `assignAddress` [doesn't set up an IPv6 address](https://github.com/moby/moby/blob/330d777c53fbbf734178d6a35c9dc0a5070ba4ac/libnetwork/network.go#L1182) in the endpoint, even if one is required - giving this driver code a chance to invent one. After the driver's endpoint has been created, the driver-invented address is reserved by a second call to `assignAddress`.

The flag passed to `NetworkOptionDeferIPv6Alloc` is the result of a check that there are [enough bits in `fixed-cidr-v6`](https://github.com/moby/moby/blob/330d777c53fbbf734178d6a35c9dc0a5070ba4ac/daemon/daemon_unix.go#L1014-L1021) for the driver to generate the address.

So, before this change, in an IPv6 enabled network - if there wasn't already an IPv6 address it was because the daemon had arranged it that way.

Now, if there's no IPv6 address, it may be because the daemon is hoping the driver will invent an address for the default bridge - or it may be because the container doesn't support IPv6. In the latter case, there may be fewer than 48 host-bits, and that's not a problem.

Reply:
> If this is a special case for the default bridge network, should we fail loudly when those conditions are not met, e.g. ones > 80 && !n.postIPv6?

The network driver doesn't know about 'postIPv6'. Checking it in `libnetwork.Network` would mean libnetwork understanding when and why the option needs to be set - but that's the daemon's job.

I'll replace the comment with your equivalent text.